api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java

```diff
@@ -0,0 +1,139 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+
+/**
+ * Offline cluster API. Handles cronjobs for cleanup and upgrade of older
+ * clusters. Methods here should be access restricted as, unlike
+ * ClusterController, these endpoints run as the Workbench service account.
+ */
+@RestController
+public class OfflineClusterController implements OfflineClusterApiDelegate {
+  private static final Logger log = Logger.getLogger(OfflineClusterController.class.getName());
+
+  private final Provider<ClusterApi> clusterApiProvider;
+  private final Provider<WorkbenchConfig> configProvider;
+  private final Clock clock;
+
+  @Autowired
+  OfflineClusterController(
+      @Qualifier(NotebooksConfig.SERVICE_CLUSTER_API) Provider<ClusterApi> clusterApiProvider,
+      Provider<WorkbenchConfig> configProvider,
+      Clock clock) {
+    this.clock = clock;
+    this.configProvider = configProvider;
+    this.clusterApiProvider = clusterApiProvider;
+  }
+
+  /**
+   * checkClusters deletes older clusters in order to force an upgrade on the
+   * next researcher login. This method is meant to be restricted to invocation
+   * by App Engine cron.
+   *
+   * The cluster deletion policy here aims to strike a balance between enforcing
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. Since a cluster was last accessed, its spent over half its remaining
+   *    lifespan idle (as defined by [1]). This dynamic limit means that as the
+   *    upgrade horizon approaches, we'll be more aggressive in upgrading.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
+    WorkbenchConfig config = configProvider.get();
+    Instant now = clock.instant();
+    Duration maxAge = Duration.ofDays(config.firecloud.clusterMaxAgeDays);
+
+    ClusterApi clusterApi = clusterApiProvider.get();
+    List<Cluster> clusters;
+    try {
+      clusters = clusterApi.listClusters(null, false);
+    } catch (ApiException e) {
+      throw ExceptionUtils.convertNotebookException(e);
+    }
+
+    int errors = 0;
+    int activeDeletes = 0;
+    int unusedDeletes = 0;
+    for (Cluster c : clusters) {
+      String clusterId = c.getGoogleProject() + "/" + c.getClusterName();
+      if (ClusterStatus.UNKNOWN.equals(c.getStatus()) || c.getStatus() == null) {
+        log.warning(String.format("unknown cluster status for cluster '%s'", clusterId));
+        continue;
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
```

Comment:
What else will we handle in future?

Is there a status here for "paused" that's different from STOPPED? I think we'd want to delete clusters when they are paused...

Reply:
Stopped is (I believe) the same as paused.

Reply:
OK. 

It seems to me that the right policy is to feel free to delete clusters in a paused / STOPPED state whenever we think is reasonable, but deleting clusters when they are RUNNING just because they are somewhat old is not very nice... we may just happen to be running our cron job in the middle of your lengthy computation.

Perhaps we should have two day settings: # of days after we delete a paused cluster, and # of days after we delete a running cluster. The latter would be significantly bigger... this should give us several days of runs to attempt to kill the cluster when it was paused, instead of running.

Reply:
👍 to not deleting running clusters.

Reply:
Also note: there is no autopausing yet, so I'm simulating this signal by looking at the lastAccessed time; later we can switch this to look at `STOPPED`